packages/blocks/github-pr-overview/src/app.tsx

```diff
@@ -0,0 +1,214 @@
+      githubEntityTypeIds !== undefined
+    ) {
+      collectPrsAndSetState(
+        githubEntityTypeIds[GITHUB_ENTITY_TYPES.PullRequest],
+        aggregateEntities,
+        accountId,
+        5,
+        setAllPrs,
+      );
+    }
+  }, [
+    githubEntityTypeIds,
+    selectedPullRequestId,
+    aggregateEntities,
+    accountId,
+    setAllPrs,
+  ]);
+
+  // Block has been initialized with a selected Pull Request, get associated info
+  React.useEffect(() => {
+    if (
+      selectedPullRequestId !== undefined &&
+      githubEntityTypeIds !== undefined
+    ) {
+      void getPrs(
+        githubEntityTypeIds[GITHUB_ENTITY_TYPES.PullRequest],
+        aggregateEntities,
+        accountId,
+        1,
+        selectedPullRequestId,
+      ).then((pullRequests) => {
+        if (pullRequests) {
+          const pr = pullRequests.results[0];
+          setPullRequest(pr);
+        }
+      });
+    }
+  }, [
+    githubEntityTypeIds,
+    selectedPullRequestId,
+    aggregateEntities,
+    accountId,
+    setPullRequest,
+  ]);
+
+  React.useEffect(() => {
+    if (
+      selectedPullRequestId !== undefined &&
+      githubEntityTypeIds !== undefined
+    ) {
+      collectReviewsAndSetState(
+        githubEntityTypeIds[GITHUB_ENTITY_TYPES.Review],
+        aggregateEntities,
+        accountId,
+        1,
+        setReviews,
+        selectedPullRequestId,
+      );
+    }
+  }, [
+    githubEntityTypeIds,
+    selectedPullRequestId,
+    aggregateEntities,
+    accountId,
+    setReviews,
+  ]);
+  React.useEffect(() => {
+    if (
+      selectedPullRequestId !== undefined &&
+      githubEntityTypeIds !== undefined
+    ) {
+      collectPrEventsAndSetState(
+        githubEntityTypeIds[GITHUB_ENTITY_TYPES.IssueEvent],
+        aggregateEntities,
+        accountId,
+        1,
+        setEvents,
+        selectedPullRequestId,
+      );
+    }
+  }, [
+    githubEntityTypeIds,
+    selectedPullRequestId,
+    aggregateEntities,
+    accountId,
+    setEvents,
+  ]);
+
+  let props: GithubPrOverviewProps | undefined;
+
+  /** @todo - handle missing data */
+  if (
+    pullRequest !== undefined &&
+    reviews !== undefined &&
+    events !== undefined
+  ) {
+    props = {
+      pullRequest,
+      reviews,
+      events,
+      setSelectedPullRequestId: setSelectedPullRequestIdAndPersist,
+    };
+  }
+
+  /** @todo - Filterable list to select a pull-request */
+  return (
+    <div>
+      {selectedPullRequestId && isDefined(props) ? (
+        <GithubPrOverview
+          pullRequest={props.pullRequest}
+          reviews={props.reviews}
+          events={props.events}
+          setSelectedPullRequestId={props.setSelectedPullRequestId}
+        />
+      ) : selectedPullRequestId === undefined && allPrs.size > 0 ? (
+        <PullRequestSelector
+          setSelectedPullRequestId={setSelectedPullRequestIdAndPersist}
+          allPrs={allPrs}
+        />
+      ) : (
+        <CircularProgress />
```

Comment:
At the moment a user will see an endless spinner if the right types aren't available or no PRs are available.

I would introduce an additional piece of state, `error/failureMessage`, which you can set at the different points at which the process can go wrong, and display it if it's defined.

Reply:
I've added a `BlockState` enum, although at the moment there's just an `Error` state without an error message. 

Adding the message is probably worth-while but I think we can do that in our next set of iteration on the block, I've at least cleaned up the logic for the various states and it should be easier to slot things in and reason about them.